crates/turborepo-telemetry/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+//! Turborepo's telemetry library. Handles sending anonymous telemetry events to
+//! the Vercel API in the background.
+//!
+//! More detail is available at https://turbo.build/repo/docs/telemetry.
+
+#![feature(error_generic_member_access)]
+
+pub mod client;
+pub mod config;
+pub mod errors;
+pub mod events;
+pub mod telemetry;
+
+use std::time::Duration;
+
+use config::{ConfigError, TelemetryConfig};
+use events::TelemetryEvent;
+use futures::{stream::FuturesUnordered, StreamExt};
+use once_cell::sync::OnceCell;
+use thiserror::Error;
+use tokio::{
+    select,
+    sync::{mpsc, oneshot},
+    task::{JoinError, JoinHandle},
+};
+use tracing::{debug, error};
+use turborepo_ui::{BOLD, GREY, UI};
+use uuid::Uuid;
+
+const BUFFER_THRESHOLD: usize = 10;
+
+static EVENT_TIMEOUT: Duration = Duration::from_millis(1000);
+static NO_TIMEOUT: Duration = Duration::from_secs(24 * 60 * 60);
+static REQUEST_TIMEOUT: Duration = Duration::from_secs(10);
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Failed to initialize telemetry")]
+    InitError(#[from] ConfigError),
+    #[error("Failed to send telemetry event")]
+    SendError(#[from] mpsc::error::SendError<TelemetryEvent>),
+    #[error("Failed to record telemetry")]
+    Join(#[from] JoinError),
+    #[error("Telemetry already initialized")]
+    AlreadyInitialized(),
+}
+
+pub type TelemetrySender = mpsc::UnboundedSender<TelemetryEvent>;
+
+/// The handle on the `Worker` tokio thread, along with a channel
+/// to indicate to the thread that it should shut down.
+pub struct TelemetryHandle {
+    exit_ch: oneshot::Receiver<()>,
+    handle: JoinHandle<()>,
+}
+
+static SENDER_INSTANCE: OnceCell<TelemetrySender> = OnceCell::new();
+
+// A global instance of the TelemetrySender.
+pub fn telem(event: events::TelemetryEvent) {
+    let sender = SENDER_INSTANCE.get();
+    match sender {
+        Some(s) => {
+            let result = s.send(event);
+            if let Err(err) = result {
+                debug!("failed to send telemetry event. error: {}", err)
+            }
+        }
+        None => {
+            if cfg!(debug_assertions) {
+                error!("[DEVELOPMENT ERROR] telemetry sender not initialized");
```

Comment:
Any reason we're not including this message in the panic call?

Reply:
I wasn't seeing this surfaced when I would hit the panic so I added an `error`, but moved to the panic instead!